### ISSUE TITLE
Feature request: rewriting without eta

### ISSUE THREAD

**Issue by mikeshulman** [agda/agda#5961]
I think it would be nice to have a version of `--rewriting` that does pure syntactic matching rather than reasoning up to eta-equality.

This would mean that ordinary confluence is insufficient to guarantee completeness of the conversion checker with respect to definitional equality.  However, if I understood the [partial resolution](https://github.com/agda/agda/issues/5923#issuecomment-1140434684) of #5923 correctly, this is also the case for rewriting _with_ eta-equality, because of singleton types.  (Please correct me if that's wrong.)

It was [suggested](https://github.com/agda/agda/issues/5923#issuecomment-1140844014) at #5923 that there could be some kind of "typed confluence check" that could suffice to ensure completeness.  That sounds interesting, though potentially a substantial amount of theoretical work, but I think it would be valuable to have such a mode even before the theoretical work is done.

Broadly speaking, the reason I think this would be nice is that Agda with rewriting can function as a sort of logical framework for "shallowly" implementing and reasoning *in* (as opposed to reasoning *about*) new type theories.  The role of rewriting is then essentially to be the "beta-reduction" of the object-language.  I think this is a slightly different point of view on rewriting than when it is viewed as making the existing type theory more convenient by making certain equalities that already hold propositionally into definitional ones.  In particular, because beta-reduction doesn't generally reason up to eta-equality (that would produce infinite loops), this use of rewriting would work better with pure syntactic matching.  (This perspective also suggests that it would be nice for the user to be able to add new eta-rules in addition to new beta-rules, pushing Agda in the direction of [Andromeda](https://arxiv.org/pdf/2103.07397.pdf).)

Obviously completeness of conversion-checking, and associated issues like subject reduction, are important properties not just in theory but in practice.  However, in this sort of application, I have the sense that good metatheoretic properties like this can't always be expected to hold globally in the framework, but only under certain restrictions on the context (or other aspects of the theory), such as a [worlds declaration](http://twelf.org/wiki/%25worlds) in Twelf.  That's certainly the case in my current application: subject reduction fails globally, but I expect that it holds in a restricted world.  So from this point of view, the fact that eta-free rewriting would break these properties globally isn't unexpected and doesn't vitiate its purpose.

**Comment by jespercockx** [comment]
> However, if I understood the https://github.com/agda/agda/issues/5923#issuecomment-1140434684 of https://github.com/agda/agda/issues/5923 correctly, this is also the case for rewriting with eta-equality, because of singleton types. (Please correct me if that's wrong.)

Indeed, this is what https://github.com/agda/agda/issues/5929 is about. There's other open issues regarding rewriting and eta-equality, see for example https://github.com/agda/agda/issues/3871, https://github.com/agda/agda/issues/4755, https://github.com/agda/agda/issues/5174, https://github.com/agda/agda/issues/5929. I believe that these might be easier to fix with "plain" rewrite rules (that don't take eta into account).

> (This perspective also suggests that it would be nice for the user to be able to add new eta-rules in addition to new beta-rules, pushing Agda in the direction of [Andromeda](https://arxiv.org/pdf/2103.07397.pdf).)

See https://github.com/agda/agda/issues/4319. I've made an attempt to implement this once, but I got discouraged by the chain of dependencies of things that would need to be changed in Agda in order to support this properly. In particular, evaluation of functions by pattern matching needs to respect user-defined eta/extensionality rules, so matching needs to become type-aware as well. Though I guess the alternative solution would be to *not* make them type-aware, and instead add some kind of restriction that makes it impossible to pattern match on datatypes with custom eta rules. I'll have to think about that some more...
